File: app/src/com/ternaryop/phototumblrshare/db/DBHelper.java
package com.ternaryop.phototumblrshare.db;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

public class DBHelper extends SQLiteOpenHelper {
	public static final String DB_NAME = "phototumblrshare.db";
 	private static final int SCHEMA_VERSION = 4;

 	private static DBHelper instance = null;

 	private LastPublishedPostCacheDAO lastPublishedPostCacheDAO;
 	private PostTagDAO postTagDAO;
 	private BirthdayDAO birthdayDAO;
 	 
	public static DBHelper getInstance(Context context) {

		// Use the application context, which will ensure that you
		// don't accidentally leak an Activity's context.
		// See this article for more information: http://bit.ly/6LRzfx
		if (instance == null) {
			instance = new DBHelper(
					context.getApplicationContext());
		}
		return instance;
	}

	/**
	 * Constructor should be private to prevent direct instantiation. make call
	 * to static factory method "getInstance()" instead.
	 */
	private DBHelper(Context context) {
		super(context, DB_NAME, null, SCHEMA_VERSION);
		lastPublishedPostCacheDAO = new LastPublishedPostCacheDAO(this);
		postTagDAO = new PostTagDAO(this);
		birthdayDAO = new BirthdayDAO(this);
	}

	@Override
	public void onCreate(SQLiteDatabase db) {
		lastPublishedPostCacheDAO.onCreate(db);
		postTagDAO.onCreate(db);
		birthdayDAO.onCreate(db);
	}

	@Override
	public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
		lastPublishedPostCacheDAO.onUpgrade(db, oldVersion, newVersion);
		postTagDAO.onUpgrade(db, oldVersion, newVersion);
		birthdayDAO.onUpgrade(db, oldVersion, newVersion);
    }

	public LastPublishedPostCacheDAO getLastPublishedPostCacheDAO() {
		return lastPublishedPostCacheDAO;
	}
	
	public PostTagDAO getPostTagDAO() {
		return postTagDAO;
	}
	
	public BirthdayDAO getBirthdayDAO() {
		return birthdayDAO;
	}
}

File: db/src/main/java/com/ternaryop/photoshelf/db/DBHelper.kt
package com.ternaryop.photoshelf.db

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

/**
 * Constructor should be private to prevent direct instantiation. make call
 * to static factory method "getInstance()" instead.
 */
class DBHelper private constructor(context: Context) : SQLiteOpenHelper(context, DB_NAME, null, SCHEMA_VERSION) {

    val tagMatcherDAO = TagMatcherDAO(this)
    val tumblrPostCacheDAO = TumblrPostCacheDAO(this)

    override fun onCreate(db: SQLiteDatabase) {
        tumblrPostCacheDAO.onCreate(db)
        tagMatcherDAO.onCreate(db)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        tumblrPostCacheDAO.onUpgrade(db, oldVersion, newVersion)
        tagMatcherDAO.onUpgrade(db, oldVersion, newVersion)
    }

    override fun onConfigure(db: SQLiteDatabase) {
        db.setForeignKeyConstraintsEnabled(true)
    }

    companion object {
        private const val DB_NAME = "photoshelf.db"
        private const val SCHEMA_VERSION = 3

        private var instance: DBHelper? = null

        fun getInstance(context: Context): DBHelper {
            val currentInstance = instance

            if (currentInstance != null) {
                return currentInstance
            }

            return synchronized(DBHelper::class.java) {
                var newInstance = instance
                if (newInstance == null) {
                    newInstance = DBHelper(context.applicationContext)
                    instance = newInstance
                }
                newInstance
            }
        }
    }
}
